### ISSUE TITLE
Feature request/comments superlist

### ISSUE THREAD

**Issue by MisterVulcan** [tridesclous/tridesclous#57]
Okay, here's a preliminary list of features/improvements I'd like to see to `tdc`. Please forgive my presumptuousness and criticism in the case they're unwarranted. I'll try and revise it as I continue to use the API and the UI and get better acquainted, but for now this is a start:

**Essentials** first
API Level
1. More comprehensive documentation. This is probably a bit obvious, but it really needs doing so users don't have to dig into the source code when they need to use the object classes for something. This is at least partially done through the jupyter/python script examples, but it needs to be more complete.
 
2. Standardization of error message language. Some of the exceptions are in English, others are in French (e.g. `peeler` and `sawchaincut` sorting algorithm). As a native English speaker I am willing to help with this once I have more time. Ditto for the documentation, some things can be phrased better than they are currently.

 3. More to come...

UI Level
1. Handled exceptions for UI errors. This is a big one; I understand how for the sake of getting functions implemented it's easier to abort the interface when an exception occurs, and you probably have other good reasons design-wise for doing this. But, much like the API documentation, it's an annoyance to deal with. Some of this is the `qt5` framework's fault, and there are a bunch of exceptions it triggers that could likely be made to fail silently without harming anything.

2. Catalogue version control. As I understand it, the only way to save/recover multiple revisions of a catalogue is to create savepoints in `cataloguewindow` and then manually rename catalogue folders in the project directory. The `phy` framework works better in this regard, although their catalogues have a more cryptic directory organization spread over multiple hidden folders. Implementing a sort of 'version control' for catalogues may take time to do properly, but it would be worth it.

3. Related to (2), `cataloguewindow` really needs an undo/redo function for merging/splitting clusters. Another thing that `phy` does well that could be improved upon in `tdc`.

4. Continuity for catalogue initialization parameters. There's a lot of variation in the quality of the recordings I use `tdc` to analyze, so I frequently have to re-run the peak detection/cluster cutting algorithms after tweaking the settings (e.g. the `min/max relative thresholds`). Some method to keep track of the parameters used for the previous catalogue initialization, instead of automatically populating the settings forms with default values, would be welcome. Maybe a button to reset the init. parameters, but otherwise keep those that were previously used.

5. Option to manually change cell label for a specific cluster, for both `pairlist` & `clusterlist`. Self-explanatory, and completely necessary when dealing with more than 2 clusters with a high similarity.

6. Option to automatically `re-label by rms` and/or `compute metrics` after merging/splitting clusters. This is related to (5) and would improve manual curation workflow efficiency for those of us with the computer horsepower to handle it.

7. Navigation in `ndscatter` window. Unless I'm wrong, you've locked the `pyqtgraph` window settings for the PCA view for stability's sake, however there isn't a way to pan/zoom the same way as other viewer windows. This is problematic, and based on my past experience with `pyqtgraph` shouldn't be a huge deal to fix.

**Nitpicks** second (in no particular order of importance)
1. File import wizard for legacy `.plx` files is broken. Latest version of `python-neo` works with old plexon files (I come from a very poor lab), and I have no problems importing `.plx` files into `tdc` from a python interpreter - however this feature is broken in the UI for some reason. This is a nitpick because I had a workaround for this before switching to `tdc`, but it would ultimately save a few hours and GB's of storage space if I didn't have to convert my recordings to a binary format before importing them into `tdc`, not to mention needing to manually specify the A/D rate and # of channels. Definitely willing to help with this issue.

2. Better integration with `ephyviewer`. It looks like you're already toying with the idea of integrating `ephyviewer` into the `tdc` workflow, but it would be nice to be able to view any of the catalogue segments before filtering/analysis instead of just the first one in the list.

3. Fix for `qt5` `cataloguewindow` interface (un)docking. As it is, there appears to be a limitation on the maximum window sizes when a subview (e.g. `waveformviewer`) is undocked from the main `cataloguewindow`. For instance, I have a 2-monitor setup with different resolutions, and if I undock `waveformviewer` and move it to a different monitor, the new window's dimensions are still tied to the parent window's and cannot be adjusted freely. Probably a default setting somewhere in `qt5` that could be changed.

4. Plot `unclassified` and `alien` clusters in `ndscatter` viewer. Self-explanatory, and would be nice for those of us that are control freaks when it comes to curating the automated clustering results. Yes, I can just select all of the peaks in `alien` and move them to a new cluster, but it seems like an unnecessary extra step.

5. Option to sort by: `cell label` in `clusterlist`. Would save me a lot of scrolling when I manually curate my clusters.

6. Remove the `sort by` drop-down form in `clusterlist` and instead tie sorting priority to the last-clicked column header. This would reduce the interface clutter, and I'm pretty sure it's already implemented in `qt5`. As an example, click on the `cluster_label` header to re-arrange `clusterlist` instead of having to select `cluster_list` from the drop-down form.

7. Add a `mua` cluster label in addition to the `trash` label. Another feature of `phy` that should be included.

8. Option to include `unclassified, alien, noise, trash` clusters in `peeler` analysis/catalogue. I do place cell analysis, which sometimes means I want to see _all_ of the detected peaks and the heatmaps created from them, including the bad/unclassified ones. When I extract spike clusters and timestamps from a `CatalogueConstructor` object after using `peeler`, it appears that only the `trash` and "good" clusters are included. Unless I'm missing something, this would be a welcome additional feature.

9. Option to export clusters, spike times directly from `cataloguewindow` or `peelerwindow`. I can do this in python with the `CatalogueConstructor` class, but it'd be nice to have the option for people that need the clustering data for further analysis but don't want to delve into the dark realms of API's. `numpy` has a nice `.csv` export feature that would work well.

**Wishlist** last (these would be nice, but are **not** essential)
1. Free-form 3D rotation of `ndscatter` view. Yes, you already have this *mostly* done with the `random tour` feature, but it could be a lot better. I'm not sure that `pyqtgraph` fully supports 3d scatter views, so it might be necessary to use a different library. Something [like this](http://projector.tensorflow.org/) would be super-cool, although I'm not sure if it would be worth the trouble to implement.

2. Multiple processing threads for `qt5` interface. I know from personal experience that `qt5` multi-threading is tricky business, but `cataloguewindow` would be _so much better_ if it were possible. Again, no idea if `pyqtgraph` would even support this.

3. `tensorflow` integration for massively-parallel clustering algorithms. There are lots of other possibilities, but just a cursory glance at its feature set suggests that `tensorflow` would be _closest_ to a drop-in substitution for `sklearn`. Just think about the possibilities...you could give `klustakwik` and `kilosort` a run for their money, as we say in the USA. Something I'd be willing to help with, as I've been looking for an excuse to learn `tensorflow` for a while now.

I hope that these suggestions/criticisms have been primarily constructive, and look forward to continuing to use `tdc` and perhaps contributing to its improvement. I like the direction it's heading versus `phy`, especially the inclusion of 3rd-party in addition to homebrewed clustering algorithms. By no means do I expect to see all or most of these suggestions implemented, but hopefully it helps give you some insight into how to improve `tdc` so that other labs can utilize it for their needs. Please keep this project under active development if possible - there aren't too many open-sourced cluster sorting tools available, and even fewer of them continue to be updated. Hopefully the comprehensiveness of this list belies my enthusiasm for the platform and desire to see it succeed. Ciao!

**Comment by samuelgarcia** [comment]
Dear Mr Vulcan,
I am very that you spend so much time in playing/analysing TDC.
I think you are the very first serious user outside my colleague and knowledge.

I am agree with lot (almost?) your comments and sugestions.
It appear that you are also a dev, so I guess you will be able to help for implementing all theses.

For **API level** 100% agree. Crutial portion of code are still drafty. Too bad.
I promise a good clean and a doc for all methods.
I plan to do it just before a first release.
I am working in making the peeler very fast with sparse template (and pythran or opencl).
I will clean and doc it just after this peeler improvement.

For all other item : could make one issue for one wish. So I will be able to discuss them independently and I (or you) will be able to close them one by one with some PR.

All of theses list indeed contructive. I will try my best to make patches.
I will try to keep TDC alive as long as possible.
I have a permanent position in my lab so it is easier to consider long term project.

Best,

**Comment by MisterVulcan** [comment]
Thank you! Glad to be of help. I'll open a few issues to start with, but I'd like to keep this one open as a sort of reference, unless you were already planning on posting some project milestones and so forth. Ah, tenure, must be nice...

**Comment by samuelgarcia** [comment]
keep this one open no problem

**Comment by samuelgarcia** [comment]
Done:
  * UI Level : 1
  * UI Level : 4


For *Nitpicks:9* (export the results), do you have clear ideas on this.
the problem is more diffcult than it appear because:
  * it is possible to have several channel groups, so we can have the same
    'cluster_label' for diffrent chan_grp
  * it is possible to have several segment for a given dataset. And so the spike_index res tart at 0 for 
    each
  * Depending on the format, with neo, the t_start of each segment can be 0s or an abcolut time
  * The Peeler tag each peak to a given 'cluster_label' but ion the catalogue you can group to same cell
   several template so cluster have also 'cell_label'. This is cover the case when 2 template in a burst 
   represent the same neuron.

I plan to code easy export at API and UI level but I have some question:
  * which format : csv, binary, mat, hdf5 or all of them ?
  * spike index or spike time ?
  * 'cluster_label' (easy) or 'cell_label' (need to convert)
  * one file per segment ?
  * on file per chan_grp ?
  * one file per cluster ?
  * clear file nameing ?

What is you user point of view on this ?

**Comment by samuelgarcia** [comment]
Some of this super lis is in #71:
  * Handled exceptions for UI errors
  * Continuity for catalogue initialization parameters.
   *Option to export clusters from MainWindow

They are still things to do.
I was working on the real time part of TDC (see demo_online).

Could you send me the PLX buggy file ?

**Comment by MisterVulcan** [comment]
Hi, sorry I haven't been actively helping with this issue for a while; had real life stuff to deal with. I'll try and pull the latest release this weekend and get back to you on the .plx issue - will edit this comment when I do.

**Comment by samuelgarcia** [comment]
Note that neo have been release 2 day ago, maybe it has fixed teh plx problem.
I don't known. In case not, send me the file.
